File: Processing/LED_Editor/Checkbox.java
import processing.core.*;
import java.lang.Math;

public class Checkbox
{
	// colors
	public static int BACKGROUND_OFF = 0xFF660000;
	public static int BACKGROUND_ON  = 0xFFFF3333;
	public static int TEXT_OFF       = 0xFFFFFFFF;
	public static int TEXT_ON        = 0xFF000000;
	public static int STROKE_NORMAL  = 0xFF000000;
	public static int STROKE_HOVER   = 0xFFFFFFFF;

	private String  label = "";
	private float   posX  = 0;
	private float   posY  = 0;
	private float   size  = 1;
	private boolean state = false;

	public Checkbox(String label, float posX, float posY, float size, boolean state)
	{
		this.label = label;
		this.posX  = posX;
		this.posY  = posY;
		this.size  = size;
		this.state = state;
	}

	public int getPosX(int dx, int f)
	{
		return Math.round(dx + f * posX);
	}

	public int getPosY(int dy, int f)
	{
		return Math.round(dy + f * posY);
	}

	public int getSize(int f)
	{
		return Math.round(f * size);
	}

	public void setState(boolean state)
	{
		this.state = state;
	}

	public boolean getState()
	{
		return state;
	}

	public boolean isOver(int dx, int dy, int f, int mouseX, int mouseY)
	{
		int x_ = mouseX - getPosX(dx, f);
		int y_ = mouseY - getPosY(dy, f);
		return (PApplet.sq(x_) + PApplet.sq(y_) <= PApplet.sq(getSize(f) / 2));
	}

	public void draw(PGraphics g, int dx, int dy, int f)
	{
		draw(g, dx, dy, f, 0, 0, true);
	}

	public void draw(PGraphics g, int dx, int dy, int f, int mouseX, int mouseY)
	{
		draw(g, dx, dy, f, mouseX, mouseY, false);
	}

	public void draw(PGraphics g, int dx, int dy, int f, int mouseX, int mouseY, boolean icon)
	{
		g.fill(state ? BACKGROUND_ON : BACKGROUND_OFF);
		if (icon) g.noStroke();
		else g.stroke(isOver(dx, dy, f, mouseX, mouseY) ? STROKE_HOVER : STROKE_NORMAL);
		g.ellipse(getPosX(dx, f), getPosY(dy, f), getSize(f), getSize(f));

		if (!icon) {
			g.fill(state ? TEXT_ON : TEXT_OFF);
			g.text(label, getPosX(dx, f), getPosY(dy, f));
		}
	}
}

File: p5js/checkbox.js
class Checkbox {
	static BACKGROUND_OFF = '#660000';
	static BACKGROUND_ON  = '#ff3333';
	static TEXT_OFF       = '#ffffff';
	static TEXT_ON        = '#000000';
	static STROKE_NORMAL  = '#000000';
	static STROKE_HOVER   = '#ffffff';

	constructor(label, posX, posY, size, state) {
		this.label = label;
		this.posX  = posX;
		this.posY  = posY;
		this.size  = size;
		this.state = state;
	}

	getPosX(dx, f) {
		return Math.round(dx + f * this.posX);
	}

	getPosY(dy, f) {
		return Math.round(dy + f * this.posY);
	}

	getSize(f) {
		return Math.round(f * this.size);
	}

	setState(state) {
		this.state = state;
	}

	getState() {
		return this.state;
	}

	isOver(dx, dy, f, mx, my) {
		const x = mx - this.getPosX(dx, f);
		const y = my - this.getPosY(dy, f);
		const r = this.getSize(f) / 2;
		return x * x + y * y <= r * r;
	}

	// mx/my omitted => icon mode (thumbnails), full mode otherwise
	draw(dx, dy, f, mx, my) {
		const icon = mx === undefined;

		fill(this.state ? Checkbox.BACKGROUND_ON : Checkbox.BACKGROUND_OFF);
		if (icon) noStroke();
		else stroke(this.isOver(dx, dy, f, mx, my) ? Checkbox.STROKE_HOVER : Checkbox.STROKE_NORMAL);
		ellipse(this.getPosX(dx, f), this.getPosY(dy, f), this.getSize(f), this.getSize(f));

		if (!icon) {
			fill(this.state ? Checkbox.TEXT_ON : Checkbox.TEXT_OFF);
			text(this.label, this.getPosX(dx, f), this.getPosY(dy, f));
		}
	}
}
